feature/CDAP-14185 Add Google Cloud PubSub Streaming Source

rohitsinha54 opened the pull request:
### Summary
- Adds Google PubSub Streaming source.
- It uses [Apache Bahir](https://github.com/apache/bahir/tree/master/streaming-pubsub) to create DStreams.
- It creates a `SparkGCPCredentials` from the service json file if one is provided and if not then it will use [Application Default Credentials](https://cloud.google.com/docs/authentication/production) Note: ADC part needs to be tested to work yet but since plugin is realtime its pending.
- FLL is not supported for Realtime Plugins so this does not include FLL.

Issue: https://issues.cask.co/browse/CDAP-14185

albertshau commented on pom.xml:
```diff
@@ -445,6 +452,7 @@
                 org.apache.avro.mapreduce;
                 parquet.avro.*;
                 parquet.hadoop.*;
+                org.apache.spark.streaming.*;
```
It doesn't seem like this would be needed. Were you getting errors without it?

rohitsinha54 replied:
Yes. Without these I get class not found exception for org.apache.spark.streaming.pubsub.PubsubReceiver. 
I can be try being more specific to pubsub package level

albertshau commented on src/main/java/co/cask/gcp/publisher/source/GoogleSubscriber.java:
```diff
@@ -0,0 +1,109 @@
+
+
+  public static class SubscriberConfig extends GCPReferenceSourceConfig {
+    @Name("subscriber")
```
no need to have this annotation if it's the same as the variable name

albertshau commented on src/main/java/co/cask/gcp/publisher/source/GoogleSubscriber.java:
```diff
@@ -0,0 +1,109 @@
+    @Description("Cloud Pub/Sub subscriber to subscribe. If a subscriber does not exists and needs to be create then " +
+      "a topic must be specified to which the subscriber will belong.")
+    @Macro
+    private String subscriber;
```
'subscription' seems to be the correct term, based on the cloud console. Please update the reference doc and widget too.

albertshau commented on src/main/java/co/cask/gcp/publisher/source/GoogleSubscriber.java:
```diff
@@ -0,0 +1,109 @@
+
+  public static class SubscriberConfig extends GCPReferenceSourceConfig {
+    @Name("subscriber")
+    @Description("Cloud Pub/Sub subscriber to subscribe. If a subscriber does not exists and needs to be create then " +
```
Cloud Pub/Sub subscription to read from. If the subscription does not exist it will be automatically created if a topic is specified. Messages published before the subscription was created will not be read.

albertshau commented on widgets/GoogleSubscriber-streamingsource.json:
```diff
@@ -0,0 +1,65 @@
+        }
+      ]
+    }
+  ],
```
there is some non-editable schema widget that will force the schema to some value in the  UI. Check the twitter source in hydrator-plugins for an example.

rohitsinha54 replied:
Thanks.

albertshau commented on src/main/java/co/cask/gcp/publisher/source/GoogleSubscriber.java:
```diff
@@ -0,0 +1,109 @@
+    @Macro
+    private String subscriber;
+
+    @Description("Cloud Pub/Sub topic to subscribe messages from. If a topic is provided and the given subscriber " +
```
Cloud Pub/Sub topic to create the subscription from. The topic is only used if the subscription does not already exist.

rohitsinha54 replied:
I updated with the earlier message which you suggested.

albertshau commented on src/main/java/co/cask/gcp/publisher/source/GoogleSubscriber.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    JavaReceiverInputDStream<SparkPubsubMessage> pubSubMessages =
+      PubsubUtils.createStream(streamingContext.getSparkStreamingContext(), config.getProject(), config.topic,
```
what happens if the topic is not actually the topic associated with the subscription? Is it simply ignored because the subscription already exists?

rohitsinha54 replied:
Yes.

rohitsinha54 replied:
Yes.

albertshau commented on src/main/java/co/cask/gcp/publisher/source/GoogleSubscriber.java:
```diff
@@ -0,0 +1,109 @@
+    Schema.recordOf("event",
+                    Schema.Field.of("message", Schema.of(Schema.Type.STRING)),
+                    Schema.Field.of("id", Schema.of(Schema.Type.STRING)),
+                    Schema.Field.of("timestamp", Schema.of(Schema.Type.STRING))
```
you can use the timestamp logical type.

albertshau commented on src/main/java/co/cask/gcp/publisher/source/GoogleSubscriber.java:
```diff
@@ -0,0 +1,109 @@
+      StructuredRecord.builder(DEFAULT_SCHEMA)
+        .set("message", Bytes.toString(pubSubMessage.getData()))
+        .set("id", pubSubMessage.getMessageId())
+        .set("timestamp", pubSubMessage.getPublishTime()).build());
```
we should make use of the new timestamp logical types and use setTimestamp()

albertshau commented on widgets/GoogleSubscriber-streamingsource.json:
```diff
@@ -0,0 +1,65 @@
+        },
+        {
+          "widget-type": "textbox",
+          "label": "Subscriber Name",
```
Subscription

albertshau commented on widgets/GoogleSubscriber-streamingsource.json:
```diff
@@ -0,0 +1,65 @@
+        },
+        {
+          "widget-type": "textbox",
+          "label": "PubSub topic",
```
Topic

albertshau commented on widgets/GoogleSubscriber-streamingsource.json:
```diff
@@ -0,0 +1,65 @@
+          "label": "PubSub topic",
+          "name": "topic",
+          "widget-attributes" : {
+            "placeholder": "Topic name for the subscriber"
```
Topic to create the subscription from if it does not already exist

albertshau commented on widgets/GoogleSubscriber-streamingsource.json:
```diff
@@ -0,0 +1,65 @@
+          "label": "Subscriber Name",
+          "name": "subscriber",
+          "widget-attributes" : {
+            "placeholder": "Subscriber to subscribe to"
```
Subscription to read from

albertshau commented on docs/GoogleSubscriber-streamingsource.md:
```diff
@@ -0,0 +1,38 @@
+
+Description
+-----------
+This sources reads from a Google Cloud Pub/Sub topic in realtime.
```
Pub/Sub subscription

rohitsinha54 commented:
@albertshau Addressed all comments and opened https://issues.cask.co/browse/CDAP-14419 for format support. Please have a look.

The warn message changes which I discussed in person will be done another PR following right after.

rohitsinha54 commented:
Opened a jira to use DirectStream: https://issues.cask.co/browse/CDAP-14438
